### PR TITLE
lib: at_custom_cmd: introduce library and rewrite GCF SMS lib

### PR DESCRIPTION
**Pull request by eivindj-nordic**
* Fix init priority such that the library is initialized after
  nrf_modem_at_init.
* Rename the at_sms_cert library to gcf_sms_lib.
* Update gcf_sms_lib library to comply with this functionality.
* Add tests for AT custom command library and remove duplicate tests
  from gcf_sms_lib tests.
* Rename tests to match the library it is testing. Add shutdown function
  to ensure modem is shut down properly.
* Add support for multiple commands on same function call in sdk-nrf
  library instead of in libmodem. This helps keep the modem library
  more tidy.
* Remove \r\n from the start of AT command responses to make the
  format identical to the modem.
* Fix CSCA response to only return OK\r\n
* Add support for CMGF AT command when called with CMGF=0
* Make changes for cmgw expected <CR> and <ESC>.
* Let CMGW return index be zero based.
* Add support for having AT+CMMS as first command in multiple.
* Call cnmi before cmgs
* fix cmss/cmgs length
* Add support for clearing all SMS buffers using the CMGD flag "4".
* Allow unfiltered commands to be part of an command containing
  multiple AT commands, e.g. AT+CMGD=0;+CFUN=0<CR>. The unfiltered
  command cannot be the first command.
* Send a CMMS=0 to the modem before sending the last CMSS or CMGS.
  This is reqired to get the last CP-ACK from the modem.
* Update system tests so they are easier to edit.
* Change CMGD=? to return index (0-2) instead of (1-3).
* Add documentation for at_custom_cmd and gcf_sms_lib libraries.
* Add pause and resume of filters to allow sending a AT command to the
  modem in specific cases.
* Make AT filter pause and resume functions to align with AT_MONITOR
  changes.

Signed-off-by: Eivind Jølsgard <eivind.jolsgard@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/libraries/modem/at_custom_cmd.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+   :depth: 2
+
+The custom AT commands library allows the application to register AT commands with the Modem library that is handled in the application. This allows the application to add custom AT commands, and to override modem AT calls with application callbacks.
+
+To add a custom AT command or overwrite an existing command, the AT commands filter and callback must be added with the :c:macro:`AT_FILTER` macro in :file:`at_custom_cmd.h`.
+If the filter matches the start of the command sent to :c:func:`nrf_modem_at_cmd`, or the command in its entirety, the AT command is passed to the filter callback instead of to the modem.
+When returning from the callback, the content of the provided buffer is treated as the modem response.
+
+.. note::
+When the custom AT commands library is enabled, the application must not call :c:func:`nrf_modem_at_filter_set` as this will override the handler set by the library.
```
Comment (divipillai) [inline doc/nrf/libraries/modem/at_custom_cmd.rst:17]:
```suggestion
.. note::
   When the custom AT commands library is enabled, the application must not call :c:func:`nrf_modem_at_filter_set` as this will override the handler set by the library.
```

File: include/modem/at_custom_cmd.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * @brief Deinitialize custom AT commands.
+ */
+int at_custom_cmd_deinit(void);
+
+/**
+ * @brief Fill response buffer without overflowing the buffer.
+ *
+ * @note For the modem library to accept the response in @c buf as a success, the response must
+ *       contain "OK\r\n". If the response is an error, use "ERROR\r\n" or appropriate CMS/CME
```
Comment (divipillai) [inline include/modem/at_custom_cmd.h:46]:
```suggestion
 *       contain "OK\\r\\n". If the response is an error, use "ERROR\\r\\n" or appropriate CMS/CME
```
This will fix the doc build error. We have to escape the special characters.

Comment (eivindj-nordic) [reply include/modem/at_custom_cmd.h]: Thanks :)

File: tests/lib/at_custom_cmd/testcase.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+tests:
+  at_custom_cmd.functionality_test:
+    platform_allow: nrf9160dk_nrf9160_ns
```
Comment (gopiotr) [inline tests/lib/at_custom_cmd/testcase.yaml:3]: Please add integration platform option to make it possible to build this test in Jenkins CI:
```
    integration_platforms:
      - nrf9160dk_nrf9160_ns
```

File: tests/lib/gcf_sms_lib/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+tests:
+  gcf_sms_lib.functionality_test:
+    platform_allow: nrf9160dk_nrf9160_ns
```
Comment (gopiotr) [inline tests/lib/gcf_sms_lib/testcase.yaml:3]: Please add integration platform option to make it possible to build this test in Jenkins CI:
```
    integration_platforms:
      - nrf9160dk_nrf9160_ns
```

File: lib/gcf_sms_lib/Kconfig [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2022 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+menuconfig GCF_SMS_LIB
+	bool "Custom SMS commands for GCF certification"
+	depends on SOC_NRF9160_SICA
+	imply AT_CUSTOM_CMD
```
Comment (lemrey) [inline lib/gcf_sms_lib/Kconfig:9]: should be `select`, because it won't work if this is deselected

**Review by divipillai — APPROVED** [review]

File: tests/lib/at_custom_cmd/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(at_custom_cmd)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
+
+target_include_directories(app
+  PRIVATE
+  ${ZEPHYR_BASE}/../nrf/include/modem/
```
Comment (tejlmand) [inline tests/lib/at_custom_cmd/CMakeLists.txt:17]:
```suggestion
  ${ZEPHYR_NRF_MODULE_DIR}/include/modem/
```

File: tests/lib/at_custom_cmd/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Comment (tejlmand) [inline tests/lib/at_custom_cmd/CMakeLists.txt:2]: 2021 ?

File: tests/lib/gcf_sms_lib/prj.conf [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Comment (tejlmand) [inline tests/lib/gcf_sms_lib/prj.conf:2]: 2021 ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
minor nits, and please squash the doc review update into the first commit.

**Comment by eivindj-nordic** [comment]
> minor nits, and please squash the doc review update into the first commit.

Thanks, updated :)

File: tests/lib/gcf_sms_lib/prj.conf [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (c) 2022s Nordic Semiconductor ASA
```
Comment (tejlmand) [inline tests/lib/gcf_sms_lib/prj.conf:2]: `2022s` ?
```suggestion
# Copyright (c) 2022 Nordic Semiconductor ASA
```

File: tests/lib/at_custom_cmd/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(at_custom_cmd)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
+
+target_include_directories(app
+  PRIVATE
+  {ZEPHYR_NRF_MODULE_DIR}/nrf/include/modem/
```
Comment (tejlmand) [inline tests/lib/at_custom_cmd/CMakeLists.txt:17]: This becomes:
`<west-workspace>/nrf/nrf/include/modem/`

probably not your intention.
```suggestion
  {ZEPHYR_NRF_MODULE_DIR}/include/modem/
```
_

Hint: Next time, just click `Commit suggestion` to ensure you get the suggestion applied correctly :wink: 
image

**Review by tejlmand — CHANGES_REQUESTED** [review]
looking better, but still a couple of nits.

File: tests/lib/at_custom_cmd/CMakeLists.txt (hunk 10 above)
Comment (tejlmand) [reply tests/lib/at_custom_cmd/CMakeLists.txt:17]: I wonder if that `include` path is needed, as nothing seems to have failed in Jenkins.

File: lib/gcf_sms_lib/gcf_sms_lib.c [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+ */
+
+static int at_cmd_callback_cpms(char *buf, size_t len, char *at_cmd)
+{
+	int sms_buffers_used = sms_buffers_num_used_get();
+
+	/* Test */
+	if (prefix_match("AT+CPMS=?", at_cmd)) {
+		return at_custome_cmd_respond(buf, len,
+							  "+CPMS: (\"TA\"),(\"TA\")\r\nOK\r\n");
```
Comment (lemrey) [inline lib/gcf_sms_lib/gcf_sms_lib.c:231]: Indentation

File: lib/gcf_sms_lib/gcf_sms_lib.c [hunk 12]
```diff
@@ -0,0 +1,481 @@
...
+	}
+
+	/* Set */
+	if (prefix_match("AT+CPMS=\"TA\",\"TA\"", at_cmd)) {
+		return at_custome_cmd_respond(buf, len, "+CPMS: %d,%d,%d,%d\r\nOK\r\n",
+							  sms_buffers_used, SMS_BUFFER_ARRAY_SIZE,
+							  sms_buffers_used, SMS_BUFFER_ARRAY_SIZE);
```
Comment (lemrey) [inline lib/gcf_sms_lib/gcf_sms_lib.c:238]: Indentation

File: lib/gcf_sms_lib/gcf_sms_lib.c [hunk 13]
```diff
@@ -0,0 +1,481 @@
...
+	}
+	if (prefix_match("AT+CPMS=\"SM\",\"SM\",\"MT\"", at_cmd)) {
+		return at_custome_cmd_respond(buf, len,
+							  "+CPMS: \"SM\",%d,%d,"
+							  "\"SM\",%d,%d,\"MT\",%d,%d\r\nOK\r\n",
+							  sms_buffers_used, SMS_BUFFER_ARRAY_SIZE,
+							  sms_buffers_used, SMS_BUFFER_ARRAY_SIZE,
+							  sms_buffers_used, SMS_BUFFER_ARRAY_SIZE);
```
Comment (lemrey) [inline lib/gcf_sms_lib/gcf_sms_lib.c:246]: Indentation

File: include/modem/at_custom_cmd.h [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+ * @param buf Buffer to put response into.
+ * @param buf_size Size of the response buffer.
+ * @param content Response format.
+ * @param ... Format arguments.
+ *
+ * @retval 0 on success.
+ * @retval -NRF_EFAULT if no buffer provided.
+ * @retval -NRF_E2BIG if the provided buffer is too small for the response.
+ */
+int at_custome_cmd_respond(char *buf, size_t buf_size,
```
Comment (lemrey) [inline include/modem/at_custom_cmd.h:48]: `custome` -> `custom`

File: include/modem/at_custom_cmd.h [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+		const char *content, ...);
```
Comment (lemrey) [inline include/modem/at_custom_cmd.h:49]: If it's a response, consider calling it `response`

File: lib/gcf_sms_lib/gcf_sms_lib.c [hunk 16]
```diff
@@ -0,0 +1,481 @@
...
+	if (prefix_match("AT+CSCA=", at_cmd)) {
+		strncpy(sca_buff, &(at_cmd[8]), SCA_SIZE_MAX);
+		sca_buff[sizeof(sca_buff) - 1] = '\0';
+		return at_custome_cmd_respond(buf, len, "OK\r\n");
+	}
+
+	/* Read */
+	if (prefix_match("AT+CSCA?", at_cmd)) {
+		return at_custome_cmd_respond(buf, len, "+CSMS: %s\r\nOK\r\n",
+							  sca_buff);
```
Comment (lemrey) [inline lib/gcf_sms_lib/gcf_sms_lib.c:293]: Indentation

File: lib/gcf_sms_lib/gcf_sms_lib.c [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+		strncpy(sca_buff, &(at_cmd[8]), SCA_SIZE_MAX);
+		sca_buff[sizeof(sca_buff) - 1] = '\0';
+		return at_custome_cmd_respond(buf, len, "OK\r\n");
+	}
+
+	return at_custome_cmd_respond(buf, len, "ERROR\r\n");
+}
+
+/* Initialize the library. */
+int gcf_sms_lib_init(const struct device *unused)
```
Comment (lemrey) [inline lib/gcf_sms_lib/gcf_sms_lib.c:467]: I would let `SYS_INIT` be the only way to initialize this library and delete the public header entirely.

File: include/modem/gcf_sms_lib.h [hunk 18]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef GCF_SMS_LIB_H_
+#define GCF_SMS_LIB_H_
```
Comment (lemrey) [inline include/modem/gcf_sms_lib.h:8]: Let's get rid of this file.

Also, I would call this library `gfc_sms` with the `lib` part at the end, wherever it is applicable, it is redundant (`gcf_sms_lib` library vs `gcf_sms` library).

**Comment by lemrey** [comment]
Would it be possible to split the commits, one where you introduce the new library, and one where you update the existing SMS library?

File: doc/nrf/libraries/modem/at_custom_cmd.rst [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+Configuration
+*************
+
+You can enable the custom AT commands library by setting the :kconfig:option:`CONFIG_AT_CUSTOM_CMD` Kconfig option.
+
+Usage
+*****
+
+To add a custom AT command or overwrite an existing command, you must add the AT commands filter and callback with the :c:macro:`AT_CUSTOM_CMD` macro in the :file:`at_custom_cmd.h` file.
+If the filter matches the start or the entire command that is sent to the :c:func:`nrf_modem_at_cmd` function, the AT command is passed to the filter callback instead of the modem.
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/at_custom_cmd.rst:22]: "The start" is the first characters of the AT command, I'll rewrite this slightly.

**Review by lemrey — APPROVED** [review]
`at_cusom_cmd` in commit message :) other than that, looks good!

**Review by divipillai — APPROVED** [review]

File: lib/gcf_sms/gcf_sms.c [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+
+	/* Test */
+	if (prefix_match("AT+CPMS=?", at_cmd)) {
+		return at_custom_cmd_respond(buf, len, "+CPMS: (\"TA\"),(\"TA\")\r\nOK\r\n");
+	}
+
+	/* Set */
+	if (prefix_match("AT+CPMS=\"TA\",\"TA\"", at_cmd)) {
+		return at_custom_cmd_respond(buf, len,
+					      "+CPMS: %d,%d,%d,%d\r\nOK\r\n",
```
Comment (MirkoCovizzi) [inline lib/gcf_sms/gcf_sms.c:236]: Nit: there is a small misalignment. In other lines as well.

Comment (eivindj-nordic) [reply lib/gcf_sms/gcf_sms.c]: Thanks, updated.

**Review by MirkoCovizzi — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm
